src/internal.rs

```diff
@@ -138,51 +17,45 @@ impl serde::ser::Error for Error {
 /// If this returns an `Error` (other than SizeLimit), assume that the
 /// writer is in an invalid state, as writing could bail out in the middle of
 /// serializing.
-pub fn serialize_into<W, T: ?Sized, S, E>(writer: W, value: &T, size_limit: S) -> Result<()>
+pub(crate) fn serialize_into<W, T: ?Sized, O>(writer: W, value: &T, mut options: O) -> Result<()>
 where
     W: Write,
     T: serde::Serialize,
-    S: SizeLimit,
-    E: ByteOrder,
+    O: Options,
 {
-    if let Some(limit) = size_limit.limit() {
-        try!(serialized_size_bounded(value, limit).ok_or(
+    if let Some(limit) = options.limit().limit() {
+        try!(serialized_size_bounded(value, limit, &mut options).ok_or(
             ErrorKind::SizeLimit,
         ));
     }
 
-    let mut serializer = Serializer::<_, E>::new(writer);
+    let mut serializer = ::ser::Serializer::<_, O>::new(writer, options);
     serde::Serialize::serialize(value, &mut serializer)
 }
 
 /// Serializes a serializable object into a `Vec` of bytes.
 ///
 /// If the serialization would take more bytes than allowed by `size_limit`,
 /// an error is returned.
-pub fn serialize<T: ?Sized, S, E>(value: &T, size_limit: S) -> Result<Vec<u8>>
+pub(crate) fn serialize<T: ?Sized, O>(value: &T, mut options: O) -> Result<Vec<u8>>
 where
     T: serde::Serialize,
-    S: SizeLimit,
-    E: ByteOrder,
+    O: Options,
 {
-    let mut writer = match size_limit.limit() {
+    let mut writer = match options.limit().limit() {
         Some(size_limit) => {
-            let actual_size = try!(serialized_size_bounded(value, size_limit).ok_or(
+            let actual_size = try!(serialized_size_bounded(value, size_limit, &mut options).ok_or(
                 ErrorKind::SizeLimit,
             ));
             Vec::with_capacity(actual_size as usize)
         }
         None => {
-            let size = serialized_size(value) as usize;
+            let size = serialized_size(value, &mut options) as usize;
```

Comment:
Whether or not we change to Result in the public `serialized_size`, I think this should return early if `serialized_size` failed.

Reply:
Agreed.  This has been changed to short circuit